course-material/interactive-queries-and-parameterization/parameterize-sql-queries.md

```diff
@@ -1,14 +1,358 @@
+
+For example, let's define a macro to convert a value from days to years, and use this macro in our query. This macro will be called `days_to_years` and it takes as input the `column_name` and `precision`. With the default `precision` set to 2.
+
+We can obtain the number of years by dividing the number of days by 365. 
+
+We can then use the `days_to_years` macro in the query, where we are passing `age` as the input column.
+
+```{code-cell} ipython3
+%%sql --save convert
+{% macro days_to_years(column_name, precision=2) %}
+    ({{ column_name }} / 365)::numeric(16, {{ precision }})
+{% endmacro %}
+
+select
+  job, marital,
+  {{ days_to_years('age') }} as age_in_years
+from bank
+```
+
+In this query, we're converting age from days to years using our macro. The results are saved in a variable `convert`.
+
+The final rendered query can be viewed using:
+
+```{code-cell} ipython3
+final = %sqlrender convert
+print(final)
+```
+
+### Sample usage
+
+Let's use macros to handle `NULL` values. The macro below checks if the chosen column has `NULL` values.
+
+The code block defines a macro that will be used later in the SQL query. The macro is named `months_to_years_handle_null`, and its purpose is to convert the values in a specified column from months to years. Additionally, this macro is designed to handle null values in a column, hence its name. 
+
+If the value in the column is null, the macro will return `NULL`; otherwise, it will convert the value from months to years.
+
+```{code-cell} ipython3
+%%sql --save convert_duration_handle_null
+{% macro months_to_years_handle_null(column_name, precision=2) %}
+    (case when {{ column_name }} is not null then {{ column_name }} / 12 else null end)::numeric(16, {{ precision }})
+{% endmacro %}
+
+select
+  job, marital,
+  {{ months_to_years_handle_null('duration') }} as duration_in_years
+from bank
+```
+
+```{code-cell} ipython3
+final = %sqlrender convert_duration_handle_null
+print(final)
+```
+
+### Mini exercise
+
+Use the `months_to_years_handle_null` macro to iterate over the `jobs`: "blue_collar", "technician", "services", "management". For each job in the loop, calculate the average duration for the specified job. Save this query as `avg_duration`. Display the query.
+
+<!-- #region -->
+<details>
+
+<summary>Answers</summary>
+
+This modification is using the `months_to_years_handle_null` macro that you had defined earlier. This macro will convert duration from months to years for each job category, while also handling null values. The macro will apply to the average duration of each job category (where the job is equal to the current job in the loop). The output for each job category is then given an alias that indicates the job category and that it represents average duration in years.
+
+Just like before, this Jinja2 template will dynamically generate the SQL query, and the --save directive will save the query result into the avg_duration variable.
```

Comment:
Has Jinja2 been explained in a previous section? There is no mention of it besides in this instance. I'm thinking it may leave students wondering what this is referencing.

Reply:
Thank you @jpjon 
fixed